typing/typecore.ml

```diff
@@ -1941,17 +1941,64 @@ let check_univars env expans kind exp ty_expected vars =
   raise (Error (exp.exp_loc, env,
                 Less_general(kind, [Unification_trace.diff ty ty_expected])))
 
-(* Check that a type is not a function *)
-let check_application_result env statement exp =
-  let loc = exp.exp_loc in
-  match (expand_head env exp.exp_type).desc with
-  | Tarrow _ ->
-      Location.prerr_warning exp.exp_loc Warnings.Partial_application
-  | Tvar _ -> ()
-  | Tconstr (p, _, _) when Path.same p Predef.path_unit -> ()
-  | _ ->
-      if statement then
-        Location.prerr_warning loc Warnings.Statement_type
+let check_partial_application statement exp =
+  let rec f delay =
+    let ty = (expand_head exp.exp_env exp.exp_type).desc in
+    let check_statement () =
+      match ty with
+      | Tconstr (p, _, _)  when Path.same p Predef.path_unit ->
+          ()
+      | _ ->
+          if statement then
+            let rec loop {exp_loc; exp_desc; _} =
+              match exp_desc with
+              | Texp_let (_, _, e)
+              | Texp_sequence (_, e)
+              | Texp_try (e, _)
+              | Texp_ifthenelse (_, e, _)
+              | Texp_match (_, {c_rhs = e; _} :: _, _) ->
+                  loop e
+              | _ ->
```

Comment:
In the `ifthenelse` and `match` cases, `e` could a non-returning expression (`raise something`, ...), the non-unit type coming from another branch. So I'm not sure it's a good idea to recurse on them.

Reply:
Looking below, it also seems that you should add `let exception` and `let module` here if you want to be complete.

Reply:
Indeed, it's been fixed (both points).